README.md
# Cone Balls

A phantom generation package for cone beam CT geometries.

This generator analytically computes a circular cone beam projection
of a configurable number of (potentially overlapping) solid spheres of
constant density and varying random radius.

It randomly places the centers of the spheres in a cube of
configurable size. Where the spheres overlap, their linear attenuation
coefficients are summed.

The randomly generated ball positions and radii are stored for later
re-use, as well as the used cone beam geometry (as an ASTRA projection
geometry).

Cone Balls is GPU accelerated and requires CUDA.


* Free software: GNU General Public License v3
* [Documentation](https://ahendriksen.github.io/cone_balls) is currently incomplete.


## Getting Started

It takes a few steps to setup Cone Balls on your
machine. We recommend installing
[Anaconda package manager](https://www.anaconda.com/download/) for
Python 3.

### Installing with conda

Simply install with:
```
conda install -c aahendriksen -c pytorch -c astra-toolbox/label/dev -c conda-forge cone_balls
```

### Installing from source

To install Cone Balls, simply clone this GitHub
project. Go to the cloned directory and run PIP installer:
```
git clone https://github.com/ahendriksen/cone_balls.git
cd cone_balls
conda install -c astra-toolbox/label/dev -c pytorch -c conda-forge cudatoolkit=10.0 astra-toolbox pytorch torchvision tifffile
pip install -e .
```

### Usage

Cone balls is a command-line utility. It creates a directory with tiff files.

``` shell
$ cone_balls --help
Usage: cone_balls [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  bench     Time the generation of cone_balls
  generate  cone_balls generates ball phantoms for cone beam geometries

$ cone_balls generate --help
Usage: cone_balls generate [OPTIONS] DIR

  generate generates cone-beam projections of ball phantoms

  By default - 100 balls are randomly generated - 1500 projections are
  computed on a 700 x 700 detector with pixel size 1.0 x 1.0 - The source-
  object distance and the source-detector distance are 700.0, meaning   that
  the detector is centered on the origin and rotates through the object.

Options:
  --num_balls INTEGER             Number of balls to generate.
  --ball_limit INTEGER            The maximal distance from the origin of a
                                  ball
  --num_angles INTEGER            Number of angles.
  --det_row_count INTEGER         Detector row count.
  --det_col_count INTEGER         Detector column count.
  --pixel_size FLOAT              The detector pixel size.
  --SOD FLOAT                     The source object distance.
  --SDD FLOAT                     The source detector distance.
  --interactive / --no-interactive
                                  Show geometry and resulting projection
                                  images
  --ball_spec FILE
  --help                          Show this message and exit.

$ cone_balls foam --help
Usage: cone_balls foam [OPTIONS] DIR

  foam generates cone-beam projections of a foam ball phantom

  The foam ball has a radius of 0.5 and is centered on the origin. Bubbles
  can be removed from this foam phantom. The location and size of these
  bubbles can either be supplied using the --ball_spec option, or randomly
  generated.

Options:
  --num_balls INTEGER             Number of balls to generate.
  --ball_limit FLOAT              The maximal distance from the origin of a
                                  ball
  --num_angles INTEGER            Number of angles.
  --det_row_count INTEGER         Detector row count.
  --det_col_count INTEGER         Detector column count.
  --pixel_size FLOAT              The detector pixel size.
  --SOD FLOAT                     The source object distance.
  --SDD FLOAT                     The source detector distance.
  --Z FLOAT                       The Z-offset of source and detector.
  --interactive / --no-interactive
                                  Show geometry and resulting projection
                                  images
  --ball_spec FILE
  --help                          Show this message and exit.

$ cone_balls bench --help
Usage: cone_balls bench [OPTIONS]

  Time the generation of cone_balls

Options:
  --num_balls INTEGER             Number of balls to generate.
  --num_angles INTEGER            Number of angles.
  --det_pix_count INTEGER         Detector column count.
  --interactive / --no-interactive
                                  Show geometry and resulting projection
                                  images
  --help                          Show this message and exit.

```


### An example run

First generate projection data using cone_balls:
```
mkdir -p example
cone_balls generate --num_balls 10 --num_angles 50 example
```

Start a python console and run:
``` python
import astra
import astra.experimental
import numpy as np
from pathlib import Path
import tifffile
import pickle

# Load data
tiffs = [tifffile.imread(str(p)) for p in sorted(Path("example/").glob("*.tif"))]
# Put projection data in [Y, Angle, X] order
proj_data = np.array(tiffs).swapaxes(0, 1)

# Load projection geometry
pg =  pickle.load(open("example/astra_geometry.pickle", "rb"))
# Create astra projection data
p_id = astra.data3d.create('-proj3d', pg, proj_data)

vg = {'option': {'WindowMinX': -350, 'WindowMaxX': 350,
                 'WindowMinY': -350, 'WindowMaxY': 350,
                 'WindowMinZ': -350, 'WindowMaxZ': 350},
      'GridRowCount': 250,
      'GridColCount': 250,
      'GridSliceCount': 250
      }
v_id = astra.data3d.create('-vol', vg)

projector = astra.create_projector("cuda3d", pg, vg)
astra.experimental.accumulate_FDK(projector, v_id, p_id)

# Display volume data
import pyqtgraph as pq
app = pq.mkQApp()
pq.image(astra.data3d.get_shared(v_id))
app.exec_()
```

The display of data might require that you install `pyqtgraph`. This
package can be installed by executing

``` shell
conda install pyqtgraph
```

## Authors and contributors

* **Allard Hendriksen** - *Initial work*

See also the list of [contributors](https://github.com/ahendriksen/cone_balls/contributors) who participated in this project.

## How to contribute

Contributions are always welcome. Please submit pull requests against the `master` branch.

If you have any issues, questions, or remarks, then please open an issue on GitHub.

## License

This project is licensed under the GNU General Public License v3 - see the [LICENSE.md](LICENSE.md) file for details.

## Acknowledgements
* To Willem Jan Palenstijn for useful advice and discussion.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
### Fixed
### Changed
### Removed

## [0.3.1] - 2019-08-08
### Added
- Experimental support for writing out the volume

## [0.3.0] - 2019-8-07
### Added
- Python 3.7 Conda package
- Parallel beam projections
### Fixed
- Bug where output directory would not be created
### Changed
- Upgraded to pytorch 1.1.0
### Removed
- Support for cudatoolkit=9.0

## [0.2.2] - 2019-05-09
### Added
- Add --Z option to foam command to move source and detector up.
- Add example code to README.md to load cone_balls data.
### Fixed
- The --interactive option now correctly displays projections upside
  up instead of upside down.

## [0.2.1] - 2018-12-21
### Added
- foam command to generate foam bubble artifacts.

## [0.2.0] - 2018-12-21
### Added
- Add a foam command to generate foam bubble artifacts.
- Add a benchmark command
### fixed
- Improve performance by using fewer CUDA registers


## 0.1.0 - 2018-11-14
### Added
- Initial release.

[Unreleased]: https://www.github.com/ahendriksen/cone_balls/compare/v0.3.1...HEAD
[0.3.1]: https://www.github.com/ahendriksen/cone_balls/compare/v0.3.0...v0.3.1
[0.3.0]: https://www.github.com/ahendriksen/cone_balls/compare/v0.2.2...v0.3.0
[0.2.2]: https://www.github.com/ahendriksen/cone_balls/compare/v0.2.1...v0.2.2
[0.2.1]: https://www.github.com/ahendriksen/cone_balls/compare/v0.2.0...v0.2.1
[0.2.0]: https://www.github.com/ahendriksen/cone_balls/compare/v0.1.0...v0.2.0


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for cone_balls."""


tests/test_cone_balls.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `cone_balls` package."""


import unittest
import cone_balls.cone_balls as cb
import pyqtgraph as pq
import astra
import numpy as np
import torch

class TestCone_balls(unittest.TestCase):
    """Tests for `cone_balls` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_move_up(self):
        """Test something."""
        interactive = True
        pixel_size = 2.4e-3
        det_row_count = det_col_count = 500
        num_angles = 1
        sod = 10.0
        sdd = 10.0

        pg = cb.generate_cone_pg(
            (pixel_size, pixel_size), (det_row_count, det_col_count),
            num_angles, sod, sdd
        )
        ball_pos, ball_radius = cb.generate_balls(1, 1)
        ball_pos[:] = 0.0
        ball_radius[:] = 0.01

        proj_data = []
        for z in np.linspace(0, 0.5, 10):
            pg_moved = cb.move_detector(pg, z)
            pg_moved = cb.move_source(pg_moved, z)

            proj_data.append(
                list(cb.generate_projections(pg_moved, ball_pos, ball_radius))[0]
            )

        proj_data = np.array(proj_data)

        if interactive:
            app = pq.mkQApp()
            # Mark corner of 10x10 pixels, so you can check that
            # pyqtgraph puts them in the top-left corner.
            proj_data[:, :10, :10] = -1
            pq.image(proj_data, axes={"t": 0, "x": 2, "y": 1})
            app.exec_()

    def test_parallel_move(self):
        interactive = True
        pixel_size = 2.4e-3
        det_row_count = det_col_count = 500
        num_angles = 1

        pg = cb.generate_parallel_pg(
            (pixel_size, pixel_size), (det_row_count, det_col_count),
            num_angles
        )
        ball_pos, ball_radius = cb.generate_balls(1, 1)
        ball_pos[:] = 0.0
        ball_radius[:] = 0.01

        proj_data = []
        for z in np.linspace(0, 0.5, 10):
            proj_data.append(
                list(cb.generate_projections(pg, ball_pos + z, ball_radius, cone=False))[0]
            )

        proj_data = np.array(proj_data)

        if interactive:
            app = pq.mkQApp()
            # Mark corner of 10x10 pixels, so you can check that
            # pyqtgraph puts them in the top-left corner.
            proj_data[:, :10, :10] = -1e-3
            pq.image(proj_data, axes={"t": 0, "x": 2, "y": 1})
            app.exec_()

    def test_parallel(self):
        interactive = True
        pixel_size = 2.4e-3
        det_row_count = det_col_count = 500
        num_angles = 100

        pg = cb.generate_parallel_pg(
            (pixel_size, pixel_size), (det_row_count, det_col_count),
            num_angles
        )
        ball_pos, ball_radius = cb.generate_balls(100, 1)

        proj_data = np.array(list(cb.generate_projections(pg, ball_pos, ball_radius, cone=False)))

        if interactive:
            app = pq.mkQApp()
            # Mark corner of 10x10 pixels, so you can check that
            # pyqtgraph puts them in the top-left corner.
            proj_data[:, :10, :10] = -1e-3
            pq.image(proj_data, axes={"t": 0, "x": 2, "y": 1})
            app.exec_()

    def test_volume(self):

        ball_pos, ball_radius = cb.generate_balls(100, 1)

        ball_pos = torch.tensor(
            [[-1, 0, 0],
             [0, 0, 0],
             [0, 0.2, 1],
             [-1, -1, -1],
            ],
            dtype=torch.float32,
            device=torch.device('cuda'),
        )
        ball_radius = torch.tensor(
            [0.2, 0.2, 0.2, 0.2],
            dtype=torch.float32,
            device=torch.device('cuda'),
        )
        size = torch.tensor((2, 2, 2), dtype=torch.float32)
        shape = (201, 201, 201)
        voxel_size = size / torch.tensor(shape, dtype=torch.float32)
        lower_left_voxel_center = - size / 2 + voxel_size / 2
        print(voxel_size)
        print(lower_left_voxel_center)
        print(lower_left_voxel_center - voxel_size / 2)

        import cone_balls_cuda as cbc

        out_volume = torch.zeros(*shape, dtype=torch.float32).cuda()
        cbc.compute_volume(
            lower_left_voxel_center.cuda(),
            voxel_size.cuda(),
            ball_pos.cuda(),
            ball_radius.cuda(),
            out_volume,
            super_sampling=8
        )
        app = pq.mkQApp()
        pq.image(
            out_volume.detach().cpu().numpy(),
            scale=(1, -1),
            axes=dict(zip("tyx", range(3))),
        )
        app.exec_()


conda/build.sh
#!/bin/sh

echo "------------------------------------------------------------"
echo "Build Environment: "
echo "------------------------------------------------------------"

env

echo "------------------------------------------------------------"


# Inspired by  astra-toolbox/python/conda/linux_release/release.sh
if [ -z $TMP_NVCC_LOCATION ]; then
    export TMP_NVCC_LOCATION=/tmp/cuda_toolkits ;
fi

case $cudatoolkit in
    9.0)  export CUDA_URL="https://developer.nvidia.com/compute/cuda/9.0/Prod/local_installers/cuda_9.0.176_384.81_linux-run";;
    9.2)  export CUDA_URL="https://developer.nvidia.com/compute/cuda/9.2/Prod2/local_installers/cuda_9.2.148_396.37_linux";;
    10.0) export CUDA_URL="https://developer.nvidia.com/compute/cuda/10.0/Prod/local_installers/cuda_10.0.130_410.48_linux";;
    *) echo -n "cudatoolkit version $cudatoolkit not recognized. Exiting."; exit 1;;
esac

echo "------------------------------------------------------------"
echo "Installing CUDA version: $cudatoolkit"
echo "-- "

mkdir -p $TMP_NVCC_LOCATION
[ -f $TMP_NVCC_LOCATION/`basename $CUDA_URL` ] || (cd $TMP_NVCC_LOCATION; wget $CUDA_URL )

export CUDA_INSTALLER=$TMP_NVCC_LOCATION/`basename $CUDA_URL`

export CUDA_HOME=$TMP_NVCC_LOCATION/cuda-$cudatoolkit

if [ -f $CUDA_HOME/bin/nvcc ] ; then
    echo "-- CUDA already installed";
else
    echo "Installing version ${cudatoolkit} "
    echo "  with ${CUDA_INSTALLER}"
    echo "  into ${CUDA_HOME}"
    echo "  with temp space ${TMP_NVCC_LOCATION}"
    sh $CUDA_INSTALLER --silent --toolkit --toolkitpath=$CUDA_HOME --tmpdir=$TMP_NVCC_LOCATION --override
fi

echo "-- Setting CUDA_HOME: $CUDA_HOME"
echo "------------------------------------------------------------"

make clean
CUDA_HOME=$CUDA_HOME $PYTHON setup.py install --single-version-externally-managed --record record.txt || exit 1


cone_balls/cone_balls.py
# -*- coding: utf-8 -*-

"""Main module."""
import click
from tqdm import tqdm
import numpy as np
import astra
import torch
import cone_balls_cuda as cb
import pyqtgraph as pq
import tifffile
import os
import logging
import warnings
import pickle
from pathlib import Path

torch_options = {"dtype": torch.float32, "device": torch.device(type="cuda")}
BATCH = 500


def generate_vg(shape, voxel_size):
    """Generate an ASTRA volume geometry

    :param shape: (z_len, y_len, x_len)
    :param voxel_size: (z_size, y_size, x_size)
    :returns:
    :rtype:

    """
    size = np.array(shape) * np.array(voxel_size)
    min_z, min_y, min_x = - size / 2
    max_z, max_y, max_x = size / 2

    vg = {
        'option': {
            'WindowMinX': min_x,
            'WindowMaxX': max_x,
            'WindowMinY': min_y,
            'WindowMaxY': max_y,
            'WindowMinZ': min_z,
            'WindowMaxZ': max_z,
        }
    }
    vg['GridSliceCount'], vg['GridRowCount'], vg['GridColCount'] = shape
    return vg


def generate_cone_pg(det_spacing, det_shape, num_angles, SOD, SDD):
    angles = np.linspace(0, 2 * np.pi, num_angles, False)
    pg = astra.create_proj_geom(
        "cone", *det_spacing, *det_shape, angles, SOD, SDD - SOD
    )
    return astra.geom_2vec(pg)


def generate_parallel_pg(det_spacing, det_shape, num_angles):
    angles = np.linspace(0, 2 * np.pi, num_angles, False)
    pg = astra.create_proj_geom(
        "parallel3d", *det_spacing, *det_shape, angles
    )
    return astra.geom_2vec(pg)


def move_source(pg, offset):
    ret = pg.copy()
    ret["Vectors"] = np.copy(ret["Vectors"])
    src_z = ret["Vectors"][:, 2]

    src_z += offset

    return ret


def move_detector(pg, offset):
    ret = pg.copy()
    ret["Vectors"] = np.copy(ret["Vectors"])
    det_z = ret["Vectors"][:, 5]

    det_z += offset
    return ret


def generate_balls(num_balls, pos_limit):
    ball_pos = (0.5 - torch.rand(num_balls, 3, **torch_options)) * 2.0 * pos_limit
    ball_radius = torch.rand(num_balls, **torch_options) * pos_limit / 10
    return (ball_pos, ball_radius)


def generate_volume(vg, ball_pos, ball_radius, super_sampling=1):
    (min_x, max_x) = vg['option']['WindowMinX'], vg['option']['WindowMaxX']
    (min_y, max_y) = vg['option']['WindowMinY'], vg['option']['WindowMaxY']
    (min_z, max_z) = vg['option']['WindowMinZ'], vg['option']['WindowMaxZ']

    # Size and shape in Z x Y x X dimensions
    size = (max_z - min_z, max_y - min_y, max_x - min_x)
    shape = (vg['GridSliceCount'], vg['GridRowCount'], vg['GridColCount'])

    voxel_size = torch.tensor(size, **torch_options) / torch.tensor(shape, **torch_options)
    lower_left_corner = torch.tensor((min_z, min_y, min_x), **torch_options)
    lower_left_voxel_center = lower_left_corner + voxel_size / 2

    out_volume = torch.zeros(*shape, **torch_options)

    cb.compute_volume(lower_left_voxel_center, voxel_size, ball_pos,
                      ball_radius, out_volume, super_sampling=super_sampling)

    return out_volume.detach().cpu().numpy()


def generate_projections(pg, ball_pos, ball_radius, cone=True):
    # Get source position, detector position, detector u position, and
    # detector v position in Z, Y, X format ([:, ::-1]).
    ray_pos = pg["Vectors"][:, 0:3][:, ::-1]
    det_pos = pg["Vectors"][:, 3:6][:, ::-1]
    det_u = pg["Vectors"][:, 6:9][:, ::-1]
    det_v = pg["Vectors"][:, 9:12][:, ::-1]
    det_shape = (pg["DetectorRowCount"], pg["DetectorColCount"])

    ray_pos = torch.tensor(np.ascontiguousarray(ray_pos), **torch_options)
    det_pos = torch.tensor(np.ascontiguousarray(det_pos), **torch_options)
    det_u = torch.tensor(np.ascontiguousarray(det_u), **torch_options)
    det_v = torch.tensor(np.ascontiguousarray(det_v), **torch_options)

    num_angles = len(ray_pos)

    generated_projections = torch.empty(num_angles, *det_shape, dtype=torch.float32).pin_memory()
    for i in range(0, num_angles, BATCH):
        # Generate BATCH number of projection images (or n, if there
        # are fewer remaining angles)
        n = min(BATCH, num_angles - i)
        proj_data = torch.zeros(n, *det_shape, **torch_options)
        proj_data.zero_()
        idx = slice(i, i + n)
        project(
            ray_pos[idx],
            det_pos[idx],
            det_u[idx],
            det_v[idx],
            ball_pos,
            ball_radius,
            proj_data,
            cone
        )
        generated_projections[idx] = proj_data.cpu()
        del proj_data

    return generated_projections.numpy()


def project(source_pos, det_pos, det_u, det_v, ball_pos, ball_radius,
            out_proj_data, cone=True):
    cb.project_balls(
        source_pos, det_pos, det_u, det_v, ball_pos, ball_radius, out_proj_data, cone
    )
    return out_proj_data


def save_spec(dir, ball_pos, ball_radius):
    num_balls = len(ball_radius)
    spec_path = os.path.join(dir, "cone_balls_spec.txt")
    spec = np.hstack((ball_pos, np.expand_dims(ball_radius, 1)))
    header = f"""
    This file was generated with cone_balls.

    There were {num_balls} balls projected onto a cone beam geometry.

    To load this file, execute:
    ---------------------------
    >>> ball_spec = np.loadtxt(file_path)
    >>> ball_pos = ball_spec[:, :3]
    >>> ball_radius = ball_spec[:, 3]
    """
    np.savetxt(spec_path, spec, header=header)


def load_spec(spec_path):
    ball_spec = np.loadtxt(spec_path)
    ball_pos = torch.tensor(ball_spec[:, :3], **torch_options)
    ball_radius = torch.tensor(ball_spec[:, 3], **torch_options)
    return ball_pos, ball_radius


def save_geometry(dir, pg):
    geom_path = os.path.join(dir, "astra_geometry.pickle")
    with open(geom_path, "wb") as handle:
        pickle.dump(pg, handle)


def load_geometry(geometry_path):
    with open(geometry_path, "rb") as handle:
        pg = pickle.load(handle)
    return pg


def display_projections(proj_data):
    app = pq.mkQApp()
    # Flip the image vertically. Pyqtgraph displays the first row of
    # the image on top. Astra expects the first row of the image to be
    # on the bottom of the detector (in the negative direction of the
    # v vector).
    proj_data = proj_data[:, ::-1, :]
    pq.image(proj_data, axes={"t": 0, "x": 2, "y": 1},
             title="Cone balls: projection data")
    app.exec_()


def display_volume(data):
    app = pq.mkQApp()
    pq.image(
        data,
        scale=(1, -1),
        axes=dict(zip("tyx", range(3))),
    )
    app.exec_()



@click.group()
def main():
    pass


@main.command()
@click.option("--num_balls", default=100, help="Number of balls to generate.")
@click.option(
    "--ball_limit", default=200,
    help="The maximal distance from the origin of a ball"
)
@click.option("--num_angles", default=1500, help="Number of angles.")
@click.option("--det_row_count", default=700, help="Detector row count.")
@click.option("--det_col_count", default=700, help="Detector column count.")
@click.option("--pixel_size", default=1.0, help="The detector pixel size.")
@click.option("--SOD", default=700.0, help="The source object distance.")
@click.option("--SDD", default=700.0, help="The source detector distance.")
@click.option(
    "--cone/--parallel",
    default=True,
    help="Cone-beam geometry",
)
@click.option(
    "--interactive/--no-interactive",
    default=False,
    help="Show geometry and resulting projection images",
)
@click.option(
    "--ball_spec",
    default=None,
    type=click.Path(
        file_okay=True, dir_okay=False, resolve_path=True, allow_dash=False
    ),
)
@click.argument(
    "dir",
    type=click.Path(
        file_okay=False, dir_okay=True, resolve_path=True, allow_dash=False
    ),
)
def generate(
    num_balls,
    ball_limit,
    num_angles,
    det_row_count,
    det_col_count,
    pixel_size,
    sod,
    sdd,
    cone,
    interactive,
    ball_spec,
    dir,
):
    """generate generates cone-beam projections of ball phantoms

    By default
    - 100 balls are randomly generated
    - 1500 projections are computed on a 700 x 700 detector with pixel size 1.0 x 1.0
    - The source-object distance and the source-detector distance are 700.0, meaning
      that the detector is centered on the origin and rotates through the object.
    """
    dir = Path(dir).expanduser().resolve()
    click.echo(f"Writing in {dir}!")
    dir.mkdir(exist_ok=True)

    if cone:
        pg = generate_cone_pg(
            (pixel_size, pixel_size),
            (det_row_count, det_col_count),
            num_angles,
            sod,
            sdd
        )
    else:
        pg = generate_parallel_pg(
            (pixel_size, pixel_size),
            (det_row_count, det_col_count),
            num_angles
        )

    if ball_spec:
        logging.info(f"Not generating balls, using {ball_spec} file.")
        ball_pos, ball_radius = load_spec(ball_spec)
    else:
        logging.info(f"Generating {num_balls} balls.")
        ball_pos, ball_radius = generate_balls(num_balls, ball_limit)

    if interactive:
        pass  # Perhaps show geometry here in the future?

    proj_data = generate_projections(pg, ball_pos, ball_radius, cone=cone)

    if interactive:
        proj_data = np.array(proj_data, copy=False)
        display_projections(proj_data)

    if not interactive:
        # Save ball positions:
        ball_pos, ball_radius = ball_pos.cpu().numpy(), ball_radius.cpu().numpy()
        save_spec(dir, ball_pos, ball_radius)
        # Save geometry:
        save_geometry(dir, pg)
        # Save tiff stack:
        for i, p in tqdm(enumerate(proj_data)):
            filename = f"scan_{i:06d}.tif"
            path = os.path.join(dir, filename)
            tifffile.imsave(path, p)


@main.command()
@click.option("--num_balls", default=100, help="Number of balls to generate.")
@click.option(
    "--ball_limit", default=0.25,
    help="The maximal distance from the origin of a ball"
)
@click.option("--num_angles", default=100, help="Number of angles.")
@click.option("--det_row_count", default=1000, help="Detector row count.")
@click.option("--det_col_count", default=1000, help="Detector column count.")
@click.option("--pixel_size", default=1.2e-3, help="The detector pixel size.")
@click.option("--SOD", default=2.0, help="The source object distance.")
@click.option("--SDD", default=2.0, help="The source detector distance.")
@click.option("--Z", default=0.0, help="The Z-offset of source and detector.")
@click.option(
    "--cone/--parallel",
    default=True,
    help="Cone-beam geometry",
)
@click.option(
    "--interactive/--no-interactive",
    default=False,
    help="Show geometry and resulting projection images",
)
@click.option(
    "--ball_spec",
    default=None,
    type=click.Path(
        file_okay=True, dir_okay=False, resolve_path=True, allow_dash=False
    ),
)
@click.argument(
    "dir",
    type=click.Path(
        file_okay=False, dir_okay=True, resolve_path=True, allow_dash=False
    ),
)
def foam(
    num_balls,
    ball_limit,
    num_angles,
    det_row_count,
    det_col_count,
    pixel_size,
    sod,
    sdd,
    z,
    cone,
    interactive,
    ball_spec,
    dir,
):
    """foam generates cone-beam projections of a foam ball phantom

    The foam ball has a radius of 0.5 and is centered on the origin.
    Bubbles can be removed from this foam phantom.
    The location and size of these bubbles can either be supplied
    using the --ball_spec option, or randomly generated.

    """
    dir = Path(dir).expanduser().resolve()
    click.echo(f"Writing in {dir}!")
    dir.mkdir(exist_ok=True)

    if cone:
        pg = generate_cone_pg(
            (pixel_size, pixel_size),
            (det_row_count, det_col_count),
            num_angles,
            sod,
            sdd
        )
        pg = move_source(pg, z)
        pg = move_detector(pg, z)
    else:
        pg = generate_parallel_pg(
            (pixel_size, pixel_size),
            (det_row_count, det_col_count),
            num_angles
        )
        pg = move_detector(pg, z)

    if ball_spec:
        logging.info(f"Not generating balls, using {ball_spec} file.")
        ball_pos, ball_radius = load_spec(ball_spec)
    else:
        logging.info(f"Generating {num_balls} balls.")
        ball_pos, ball_radius = generate_balls(num_balls, ball_limit)

    foam_pos = torch.zeros(1, 3, **torch_options)
    foam_radius = torch.full((1,), 0.5, **torch_options)

    if interactive:
        pass  # Perhaps show geometry here in the future?

    ball_data = generate_projections(pg, ball_pos, ball_radius, cone=cone)
    foam_data = generate_projections(pg, foam_pos, foam_radius, cone=cone)

    if interactive:
        proj_data = foam_data - ball_data
        display_projections(proj_data)

    if not interactive:
        if os.path.exists(dir):
            warnings.warn(f"{dir} already exists. Overwriting files.")

        # Save ball positions:
        ball_pos, ball_radius = ball_pos.cpu().numpy(), ball_radius.cpu().numpy()
        save_spec(dir, ball_pos, ball_radius)
        # Save geometry:
        save_geometry(dir, pg)
        # Save tiff stack:
        for i, (ball, foam) in tqdm(enumerate(zip(ball_data, foam_data))):
            filename = f"scan_{i:06d}.tif"
            path = os.path.join(dir, filename)
            p = foam - ball
            tifffile.imsave(path, p)


@main.command()
@click.option("--num_balls", default=100, help="Number of balls to generate.")
@click.option(
    "--ball_limit", default=0.25,
    help="The maximal distance from the origin of a ball"
)
@click.option("--shape_z", default=1000, help="Number of slices in Z direction.")
@click.option("--shape_y", default=1000, help="Number of slices in Y direction.")
@click.option("--shape_x", default=1000, help="Number of slices in X direction.")
@click.option("--voxel_size", default=1.05e-3, help="The voxel size.")
@click.option("--super_sampling", default=1, help="By how much to oversample the voxels.")
@click.option(
    "--interactive/--no-interactive",
    default=False,
    help="Show resulting data",
)
@click.option(
    "--ball_spec",
    default=None,
    type=click.Path(
        file_okay=True, dir_okay=False, resolve_path=True, allow_dash=False
    ),
)
@click.argument(
    "dir",
    type=click.Path(
        file_okay=False, dir_okay=True, resolve_path=True, allow_dash=False
    ),
)
def foam_volume(
    num_balls,
    ball_limit,
    shape_z,
    shape_y,
    shape_x,
    voxel_size,
    super_sampling,
    interactive,
    ball_spec,
    dir,
):
    """foam generates volume data of a foam ball phantom

    The foam ball has a radius of 0.5 and is centered on the origin.
    Bubbles can be removed from this foam phantom.
    The location and size of these bubbles can either be supplied
    using the --ball_spec option, or randomly generated.

    """
    dir = Path(dir).expanduser().resolve()
    click.echo(f"Writing in {dir}!")
    dir.mkdir(exist_ok=True)

    shape = (shape_z, shape_y, shape_x)
    vg = generate_vg(shape, (voxel_size, ) * 3)

    if ball_spec:
        logging.info(f"Not generating balls, using {ball_spec} file.")
        ball_pos, ball_radius = load_spec(ball_spec)
    else:
        logging.info(f"Generating {num_balls} balls.")
        ball_pos, ball_radius = generate_balls(num_balls, ball_limit)

    foam_pos = torch.zeros(1, 3, **torch_options)
    foam_radius = torch.full((1,), 0.5, **torch_options)

    ball_data = generate_volume(vg, ball_pos, ball_radius, super_sampling=super_sampling)
    foam_data = generate_volume(vg, foam_pos, foam_radius, super_sampling=super_sampling)

    if interactive:
        proj_data = foam_data - ball_data
        display_volume(proj_data)

    if not interactive:
        if os.path.exists(dir):
            warnings.warn(f"{dir} already exists. Overwriting files.")

        # Save ball positions:
        ball_pos, ball_radius = ball_pos.cpu().numpy(), ball_radius.cpu().numpy()
        save_spec(dir, ball_pos, ball_radius)
        # Save geometry:
        save_geometry(dir, vg)
        # Save tiff stack:
        for i, (ball, foam) in tqdm(enumerate(zip(ball_data, foam_data))):
            filename = f"scan_{i:06d}.tif"
            path = os.path.join(dir, filename)
            p = foam - ball
            tifffile.imsave(path, p)


@main.command()
@click.option("--num_balls", default=100, help="Number of balls to generate.")
@click.option("--num_angles", default=1500, help="Number of angles.")
@click.option("--det_pix_count", default=700, help="Detector column count.")
@click.option(
    "--cone/--parallel",
    default=True,
    help="Cone-beam geometry",
)
@click.option(
    "--interactive/--no-interactive",
    default=False,
    help="Show geometry and resulting projection images",
)
def bench(
    num_balls,
    num_angles,
    det_pix_count,
    cone,
    interactive,
):
    """Time the generation of cone_balls
    """

    ball_limit = 200
    pixel_size = 700 / det_pix_count
    sod = 700
    sdd = 700

    if cone:
        pg = generate_cone_pg(
            (pixel_size, pixel_size),
            (det_pix_count, det_pix_count),
            num_angles,
            sod,
            sdd
        )
    else:
        pg = generate_parallel_pg(
            (pixel_size, pixel_size),
            (det_pix_count, det_pix_count),
            num_angles
        )

    logging.info(f"Generating {num_balls} balls.")
    ball_pos, ball_radius = generate_balls(num_balls, ball_limit)

    if interactive:
        pass  # Perhaps show geometry here in the future?

    proj_data = generate_projections(pg, ball_pos, ball_radius, cone=cone)

    if interactive:
        display_projections(proj_data)


if __name__ == "__main__":
    main()
